refactor(root): introducing graceful shutdown and optimise docker images

merrcury opened the pull request:
### What changed? Why was the change needed?
This PR linked to below issues
- Fixes INF-447
- Fixes  INF-443
- Fixes  INF-445


### Screenshots
New Images are now 46% Slimmer 
<img width="950" alt="Screenshot 2024-10-23 at 3 09 50 PM" src="https://github.com/user-attachments/assets/8c4616a9-f73a-4d4d-958d-ad3f606c4f15">

rifont committed:
build: make zod-to-json-schema a dynamic import

merrcury committed:
feat(application-generic): Add graceful shutdown module

Add a new module called "graceful-shutdown" to the "application-generic" library. This module includes a "ShutdownService" class that provides a timeout value for graceful shutdown. The timeout value is retrieved from the environment variable "GRACEFUL_SHUTDOWN_TIMEOUT" with a default value of 5000. This module also exports the "ShutdownService" for external use.

merrcury committed:
Merge branch 'next' of github.com:novuhq/novu into feat/docker-improvements

merrcury committed:
feat(docker): Add base Dockerfile for GHCR publishing

merrcury committed:
feat(docker): Optimize Dockerfile for GHCR publishing

merrcury committed:
feat(docker): Add .dockerignore file to exclude unnecessary files and directories

merrcury committed:
feat(docker): Add .dockerignore file and optimize Dockerfile for GHCR publishing

merrcury committed:
feat(docker): Add .dockerignore file and optimize Dockerfile for GHCR publishing

merrcury committed:
feat(docker): Add .dockerignore file and optimize Dockerfile for GHCR publishing

merrcury committed:
refactor(deps): Remove unused dependencies in package.json files

merrcury committed:
refactor(nest): Update import statement for Request and Response types in nest.controller.ts

merrcury committed:
Merge branch 'next' of github.com:novuhq/novu into feat/docker-improvements

merrcury committed:
Merge branch 'next' of github.com:novuhq/novu into feat/docker-improvements

merrcury committed:
Merge branch 'next' of github.com:novuhq/novu into feat/docker-improvements

merrcury committed:
refactor(docker): add appropriate comments to each file

rifont commented on libs/application-generic/src/index.ts:
```diff
@@ -5,14 +5,15 @@ export * from './decorators';
 export * from './dtos';
 export * from './encryption/index';
 export * from './factories/index';
+export * from './graceful-shutdown/index';
```
<!-- Chores are simple tasks that must be done before the subject can be “officially” accepted. Usually, these comments reference some common process. Try to leave a link to the process description so that the reader knows how to resolve the chore.  -->
☑ **chore:** Extra path not needed‏

```suggestion
export * from './graceful-shutdown';
```

merrcury replied:
Okay 👍🏼

merrcury replied:
Done

rifont commented on libs/application-generic/src/graceful-shutdown/shutdown.service.ts:
```diff
@@ -0,0 +1,7 @@
+@Injectable()
+export class ShutdownService {
+  public readonly timeout =
+    parseInt(process.env.GRACEFUL_SHUTDOWN_TIMEOUT, 10) || 5000;
```
<!-- Chores are simple tasks that must be done before the subject can be “officially” accepted. Usually, these comments reference some common process. Try to leave a link to the process description so that the reader knows how to resolve the chore.  -->
☑ **chore:** Please add the new environment variable to all the [env.validators](https://github.com/novuhq/novu/blob/next/apps/api/src/config/env.validators.ts) files in services that need it. This ensures that the application won't startup without the env-var being present. You will also be able to specify the `5000` default there to fully encapsulate the necessary fallback logic.‏

merrcury replied:
Done

rifont commented on packages/framework/package.json:
```diff
@@ -211,7 +211,6 @@
   "devDependencies": {
     "@apidevtools/json-schema-ref-parser": "11.6.4",
     "@arethetypeswrong/cli": "^0.16.4",
-    "@nestjs/common": "10.4.1",
```
<!-- Issues highlight specific problems with the subject under review. These problems can be user-facing or behind the scenes. It is strongly recommended to pair this comment with a suggestion. If you are not sure if a problem exists or not, consider leaving a question.  -->
⚠ **issue:** We need to keep this so that the test suites have the needed types. Please revert the removal.‏

merrcury replied:
Hey @rifont, If I am adding this into dev dependency and then running `--prod` its not getting installed and API build up is failing. any workaround/idea?

rifont replied:
That's strange, as `@novu/api` declares `@nestjs/common` as a `dependency` [here](https://github.com/novuhq/novu/pull/6754/files#diff-2c40985d6d91eed8ae85ec1c8e754a85984ee32e156a600d2b7a467423d7e338R36) which means it should be installed when using the `--prod` flag.

Could you please share the error log for further debugging?

I believe reverting the removal of this dependency is a blocker for Framework tests to pass.

merrcury replied:
Hey @rifont, Please find error below:
<img width="1236" alt="Screenshot 2024-10-29 at 11 49 50 AM" src="https://github.com/user-attachments/assets/b52ae680-c5ee-416f-83cf-75083236e081">

SokratisVidros replied:
This is now solved. See https://github.com/novuhq/novu/pull/6754/commits/4a3e98d6be620f52a25aec47a057fdd4c2033e58. 

The trick is that @novu/framework requires a range for @nestjs/common peer dependency but by pinning the @nestjs/common version via PNPM overrides we effectively teach PNPM to add the correct dependencies when installing only production dependencies.

merrcury replied:
I tested locally as well. It worked fabulous for me. Sok, we need an article for all of this 😄 .

SokratisVidros replied:
We will! 😆

SokratisVidros reviewed (approved):
😍

SokratisVidros commented on apps/api/Dockerfile:
```diff
@@ -1,72 +1,54 @@
+# Set the working directory to the API app and start the application using pm2-runtime
 WORKDIR /usr/src/app/apps/api
-CMD [ "pm2-runtime","start", "dist/main.js" ]
+CMD [ "pm2-runtime", "start", "dist/main.js" ]
```
Didn't we remove `pm2`?

merrcury replied:
Next Iteration, with ecs. We are ready to remove it but maybe in next two days

merrcury committed:
refactor(application-generic): update import statement for graceful-shutdown

merrcury committed:
Merge branch 'next' of github.com:novuhq/novu into feat/docker-improvements

merrcury committed:
Merge branch 'next' of github.com:novuhq/novu into feat/docker-improvements

merrcury committed:
Merge branch 'next' of github.com:novuhq/novu into feat/docker-improvements

merrcury committed:
feat(config): Add GRACEFUL_SHUTDOWN_TIMEOUT environment variable validation

This commit adds validation for the GRACEFUL_SHUTDOWN_TIMEOUT environment variable in multiple config files. The variable is now validated as a number with a default value of 5000. This ensures that the application gracefully shuts down within the specified timeout period.

Refs: #issue-number

merrcury committed:
Merge branch 'next' of github.com:novuhq/novu into feat/docker-improvements

SokratisVidros committed:
fix(root): Align @nestjs/common versions across the repo

The @nestjs/common global override set to "@nestjs/common@>=10.0.0 <11.0.0": "10.4.1" is necessary to teach the API application to inject it properly to @novu/framework as a peer dependency.

Refer to https://github.com/novuhq/novu/pull/6754#discussion_r1820152576

merrcury committed:
Merge branch 'feat/docker-improvements' of github.com:novuhq/novu into feat/docker-improvements

merrcury committed:
Merge branch 'next' of github.com:novuhq/novu into feat/docker-improvements

merrcury committed:
Merge branch 'next' of github.com:novuhq/novu into feat/docker-improvements
